pipelinewise/fastsync/commons/tap_postgres.py

```diff
@@ -237,23 +266,75 @@ def create_replication_slot(self):
-                self.connection_config['dbname'],
-                self.connection_config['tap_id'],
-            )
+        slot_name = self.__get_slot_name(
+            self.primary_host_conn,
+            self.connection_config['dbname'],
+            self.connection_config['tap_id'],
+        )
 
-            # Create the replication host
-            self.primary_host_query(
-                f"SELECT * FROM pg_create_logical_replication_slot('{slot_name}', 'wal2json')"
-            )
-        except Exception as exc:
-            # ERROR: replication slot already exists SQL state: 42710
-            if hasattr(exc, 'pgcode') and exc.pgcode == '42710':
-                pass
+        # Create the replication host
+        with self.primary_host_conn as conn:
+            with conn.cursor() as cur:
+                cur.create_replication_slot(slot_name, output_plugin='wal2json')
+                res = cur.fetchone()
+
+        return SlotCreationResult(
+            slot_name=res[0],
+            start_point=parse_lsn(res[1]),
+            snapshot_name=res[2],
+            plugin_name=res[3],
+        )
+
+    def get_current_lsn(self) -> int:
+        """Get the LSN of the most recent WAL availiable to the snapshotting process"""
+        # is replica_host set ?
+        if self.connection_config.get('replica_host'):
+            # Get latest applied lsn from replica_host
+            if self.version >= 100000:
+                result = self.query('SELECT pg_last_wal_replay_lsn() AS current_lsn')
+            elif self.version >= 90400:
+                result = self.query('SELECT pg_last_xlog_replay_location() AS current_lsn')
             else:
-                raise exc
+                raise Exception(
+                    'Logical replication not supported before PostgreSQL 9.4'
+                )
+        else:
+            # Get current lsn from primary host
+            if self.version >= 100000:
+                result = self.query('SELECT pg_current_wal_lsn() AS current_lsn')
+            elif self.version >= 90400:
+                result = self.query('SELECT pg_current_xlog_location() AS current_lsn')
+            else:
+                raise Exception('Logical replication not supported before PostgreSQL 9.4')
+
+        return parse_lsn(result[0].get('current_lsn'))
+
+    def get_confirmed_flush_lsn(self) -> int:
+        """
+        Get this last flushed LSN for the replication slot.
+
+        For Postgres <9.6 with defaults to the restart_lsn and confirmed_flus_lsn
+        is not availiable.
+        """
+        slot_name = self.__get_slot_name(
+            self.primary_host_conn,
+            self.connection_config['dbname'],
+            self.connection_config['tap_id'],
+        )
+
+        res = self.primary_host_query(
+            'SELECT * '
+            'FROM pg_replication_slots '
+            f"WHERE slot_name = '{slot_name}' "
+            "AND plugin = 'wal2json' "
+            "AND slot_type = 'logical'"
```

Comment:
The reason I did it this way is so that it doesn't show up in the logs with a bunch of weird indents and newlines etc. But I suspect there's a better middle ground 🤔